source/Halibut/Transport/WebSocketConnectionFactory.cs

```diff
@@ -62,15 +62,25 @@ ClientWebSocket CreateConnectedClient(ServiceEndPoint serviceEndpoint)
                         .ConfigureAwait(false).GetAwaiter().GetResult();
                 ServerCertificateInterceptor.Validate(connectionId, serviceEndpoint);
             }
+            catch
+            {
+                if (client.State == WebSocketState.Open)
+                    using (var sendCancel = new CancellationTokenSource(TimeSpan.FromSeconds(1)))
+                        client.CloseAsync(WebSocketCloseStatus.ProtocolError, "Certificate thumbprint not recognised", sendCancel.Token)
```

Comment:
Should we be using `CloseOutputAsync` here as well?

Reply:
I'm not sure to be honest. It works like this and both sides are happy. This https://stackoverflow.com/a/26744644/10784 suggests CloseAsync is a nicer way of doing things.

> There are two ways to close a WebSocket connection. The graceful way is CloseAsync which when initiated sends a message to the connected party, and waits for acknowledgement. The keyword here is sends. Remember in our previous point we discussed that you can only have a single send or receive at any given point in time? So if you are sending data, and at the same time try to CloseAsync this leads to an exception because CloseAsync will also try to send a message. The other option is to use CloseOutputAsync this is more of a “fire-and-forget” approach.